File: pos_flutter_app/lib/core/models/vehicle.dart
class VehicleBrand {
  final int id;
  final String name;
  final int typeId;
  final DateTime createdAt;

  const VehicleBrand({
    required this.id,
    required this.name,
    required this.typeId,
    required this.createdAt,
  });

  factory VehicleBrand.fromJson(Map<String, dynamic> json) {
    print('VehicleBrand.fromJson: Received JSON: $json');

    return VehicleBrand(
      id: (json['id'] as num).toInt(),
      name: json['name'] as String,
      typeId: (json['type_id'] as num).toInt(),
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'type_id': typeId,
      'created_at': createdAt.toIso8601String(),
    };
  }
}

class Vehicle {
  final int id;
  final String code;
  final int brandId;
  final String model;
  final int year;
  final String? color;
  final String? engineCapacity;
  final String? fuelType;
  final String? transmissionType;
  final String? licensePlate;
  final String? chassisNumber;
  final String? engineNumber;
  final int odometer;
  final String sourceType;
  final int? sourceId;
  final double purchasePrice;
  final String conditionStatus;
  final String status;
  final double? repairCost;
  final double? hppPrice;
  final double? sellingPrice;
  final double? soldPrice;
  final DateTime? soldDate;
  final String? notes;
  final int createdBy;
  final DateTime createdAt;
  final DateTime updatedAt;
  final VehicleBrand? brand;

  const Vehicle({
    required this.id,
    required this.code,
    required this.brandId,
    required this.model,
    required this.year,
    this.color,
    this.engineCapacity,
    this.fuelType,
    this.transmissionType,
    this.licensePlate,
    this.chassisNumber,
    this.engineNumber,
    required this.odometer,
    required this.sourceType,
    this.sourceId,
    required this.purchasePrice,
    required this.conditionStatus,
    required this.status,
    this.repairCost,
    this.hppPrice,
    this.sellingPrice,
    this.soldPrice,
    this.soldDate,
    this.notes,
    required this.createdBy,
    required this.createdAt,
    required this.updatedAt,
    this.brand,
  });

  factory Vehicle.fromJson(Map<String, dynamic> json) {
    print('Vehicle.fromJson: Received JSON: $json');

    final vehicle = Vehicle(
      id: (json['id'] as num? ?? 0).toInt(),
      code: json['code'] as String? ?? '',
      brandId: (json['brand_id'] as num? ?? 0).toInt(),
      model: json['model'] as String? ?? '',
      year: (json['year'] as num? ?? 0).toInt(),
      color: json['color'] as String?,
      engineCapacity: json['engine_capacity'] as String?,
      fuelType: json['fuel_type'] as String?,
      transmissionType: json['transmission_type'] as String?,
      licensePlate: json['license_plate'] as String?,
      chassisNumber: json['chassis_number'] as String?,
      engineNumber: json['engine_number'] as String?,
      odometer: (json['odometer'] as num?)?.toInt() ?? 0,
      sourceType: json['source_type'] as String? ?? '',
      sourceId: (json['source_id'] as num?)?.toInt(),
      purchasePrice: (json['purchase_price'] as num? ?? 0.0).toDouble(),
      conditionStatus: json['condition_status'] as String? ?? '',
      status: json['status'] as String? ?? '',
      repairCost: json['repair_cost'] != null
          ? (json['repair_cost'] as num?)?.toDouble()
          : null,
      hppPrice: json['hpp_price'] != null
          ? (json['hpp_price'] as num?)?.toDouble()
          : null,
      sellingPrice: json['selling_price'] != null
          ? (json['selling_price'] as num?)?.toDouble()
          : null,
      soldPrice: json['sold_price'] != null
          ? (json['sold_price'] as num?)?.toDouble()
          : null,
      soldDate: json['sold_date'] != null
          ? DateTime.parse(json['sold_date'] as String)
          : null,
      notes: json['notes'] as String?,
      createdBy: (json['created_by'] as num? ?? 0).toInt(),
      createdAt: DateTime.parse(
          json['created_at'] as String? ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(
          json['updated_at'] as String? ?? DateTime.now().toIso8601String()),
      brand: json['brand'] != null
          ? VehicleBrand.fromJson(json['brand'] as Map<String, dynamic>)
          : null,
    );

    print(
        'Vehicle.fromJson: Parsed vehicle: color=${vehicle.color}, fuelType=${vehicle.fuelType}, transmissionType=${vehicle.transmissionType}');
    return vehicle;
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'code': code,
      'brand_id': brandId,
      'model': model,
      'year': year,
      'color': color,
      'engine_capacity': engineCapacity,
      'fuel_type': fuelType,
      'transmission_type': transmissionType,
      'license_plate': licensePlate,
      'chassis_number': chassisNumber,
      'engine_number': engineNumber,
      'odometer': odometer,
      'source_type': sourceType,
      'source_id': sourceId,
      'purchase_price': purchasePrice,
      'condition_status': conditionStatus,
      'status': status,
      'repair_cost': repairCost,
      'hpp_price': hppPrice,
      'selling_price': sellingPrice,
      'sold_price': soldPrice,
      'sold_date': soldDate?.toIso8601String(),
      'notes': notes,
      'created_by': createdBy,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'brand': brand?.toJson(),
    };
  }

  // Getter methods
  String get brandName => brand?.name ?? 'Unknown Brand';
  String get displayName => '$brandName $model ($year)';
  String get statusDisplay => _getStatusDisplay(status);
  String get conditionDisplay => _getConditionDisplay(conditionStatus);

  String _getStatusDisplay(String status) {
    switch (status) {
      case 'available':
        return 'Tersedia';
      case 'in_repair':
        return 'Dalam Perbaikan';
      case 'sold':
        return 'Terjual';
      case 'reserved':
        return 'Dipesan';
      default:
        return status;
    }
  }

  String _getConditionDisplay(String condition) {
    switch (condition) {
      case 'excellent':
        return 'Sangat Baik';
      case 'good':
        return 'Baik';
      case 'fair':
        return 'Cukup';
      case 'poor':
        return 'Kurang';
      default:
        return condition;
    }
  }

  Vehicle copyWith({
    int? id,
    String? code,
    int? brandId,
    String? model,
    int? year,
    String? color,
    String? engineCapacity,
    String? fuelType,
    String? transmissionType,
    String? licensePlate,
    String? chassisNumber,
    String? engineNumber,
    int? odometer,
    String? sourceType,
    int? sourceId,
    double? purchasePrice,
    String? conditionStatus,
    String? status,
    double? repairCost,
    double? hppPrice,
    double? sellingPrice,
    double? soldPrice,
    DateTime? soldDate,
    String? notes,
    int? createdBy,
    DateTime? createdAt,
    DateTime? updatedAt,
    VehicleBrand? brand,
  }) {
    return Vehicle(
      id: id ?? this.id,
      code: code ?? this.code,
      brandId: brandId ?? this.brandId,
      model: model ?? this.model,
      year: year ?? this.year,
      color: color ?? this.color,
      engineCapacity: engineCapacity ?? this.engineCapacity,
      fuelType: fuelType ?? this.fuelType,
      transmissionType: transmissionType ?? this.transmissionType,
      licensePlate: licensePlate ?? this.licensePlate,
      chassisNumber: chassisNumber ?? this.chassisNumber,
      engineNumber: engineNumber ?? this.engineNumber,
      odometer: odometer ?? this.odometer,
      sourceType: sourceType ?? this.sourceType,
      sourceId: sourceId ?? this.sourceId,
      purchasePrice: purchasePrice ?? this.purchasePrice,
      conditionStatus: conditionStatus ?? this.conditionStatus,
      status: status ?? this.status,
      repairCost: repairCost ?? this.repairCost,
      hppPrice: hppPrice ?? this.hppPrice,
      sellingPrice: sellingPrice ?? this.sellingPrice,
      soldPrice: soldPrice ?? this.soldPrice,
      soldDate: soldDate ?? this.soldDate,
      notes: notes ?? this.notes,
      createdBy: createdBy ?? this.createdBy,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      brand: brand ?? this.brand,
    );
  }

  @override
  String toString() {
    return 'Vehicle(id: $id, code: $code, displayName: $displayName, status: $status)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is Vehicle && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;
}

class CreateVehicleRequest {
  final String code;
  final int brandId;
  final String model;
  final int year;
  final String color;
  final String? engineCapacity;
  final String? fuelType;
  final String? transmissionType;
  final String? licensePlate;
  final int? odometer;
  final String sourceType;
  final int? sourceId;
  final double purchasePrice;
  final String conditionStatus;
  final String? description;

  const CreateVehicleRequest({
    required this.code,
    required this.brandId,
    required this.model,
    required this.year,
    required this.color,
    this.engineCapacity,
    this.fuelType,
    this.transmissionType,
    this.licensePlate,
    this.odometer,
    required this.sourceType,
    this.sourceId,
    required this.purchasePrice,
    required this.conditionStatus,
    this.description,
  });

  Map<String, dynamic> toJson() {
    return {
      'code': code,
      'brand_id': brandId,
      'model': model,
      'year': year,
      'color': color,
      'engine_capacity': engineCapacity,
      'fuel_type': fuelType,
      'transmission_type': transmissionType,
      'license_plate': licensePlate,
      'odometer': odometer,
      'source_type': sourceType,
      'source_id': sourceId,
      'purchase_price': purchasePrice,
      'condition_status': conditionStatus,
      'description': description,
    };
  }
}

class UpdateVehicleRequest {
  final String? code;
  final int? brandId;
  final String? model;
  final int? year;
  final String? color;
  final String? engineCapacity;
  final String? fuelType;
  final String? transmissionType;
  final String? licensePlate;
  final String? chassisNumber;
  final String? engineNumber;
  final int? odometer;
  final String? sourceType;
  final int? sourceId;
  final double? purchasePrice;
  final String? conditionStatus;
  final String? status;
  final double? repairCost;
  final double? hppPrice;
  final double? sellingPrice;
  final double? soldPrice;
  final DateTime? soldDate;
  final String? notes;

  const UpdateVehicleRequest({
    this.code,
    this.brandId,
    this.model,
    this.year,
    this.color,
    this.engineCapacity,
    this.fuelType,
    this.transmissionType,
    this.licensePlate,
    this.chassisNumber,
    this.engineNumber,
    this.odometer,
    this.sourceType,
    this.sourceId,
    this.purchasePrice,
    this.conditionStatus,
    this.status,
    this.repairCost,
    this.hppPrice,
    this.sellingPrice,
    this.soldPrice,
    this.soldDate,
    this.notes,
  });

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {};

    if (code != null) data['code'] = code;
    if (brandId != null) data['brand_id'] = brandId;
    if (model != null) data['model'] = model;
    if (year != null) data['year'] = year;
    if (color != null) data['color'] = color;
    if (engineCapacity != null) data['engine_capacity'] = engineCapacity;
    if (fuelType != null) data['fuel_type'] = fuelType;
    if (transmissionType != null) data['transmission_type'] = transmissionType;
    if (licensePlate != null) data['license_plate'] = licensePlate;
    if (chassisNumber != null) data['chassis_number'] = chassisNumber;
    if (engineNumber != null) data['engine_number'] = engineNumber;
    if (odometer != null) data['odometer'] = odometer;
    if (sourceType != null) data['source_type'] = sourceType;
    if (sourceId != null) data['source_id'] = sourceId;
    if (purchasePrice != null) data['purchase_price'] = purchasePrice;
    if (conditionStatus != null) data['condition_status'] = conditionStatus;
    if (status != null) data['status'] = status;
    if (repairCost != null) data['repair_cost'] = repairCost;
    if (hppPrice != null) data['hpp_price'] = hppPrice;
    if (sellingPrice != null) data['selling_price'] = sellingPrice;
    if (soldPrice != null) data['sold_price'] = soldPrice;
    if (soldDate != null) data['sold_date'] = soldDate!.toIso8601String();
    if (notes != null) data['notes'] = notes;

    return data;
  }
}

class SetSellingPriceRequest {
  final double sellingPrice;

  const SetSellingPriceRequest({
    required this.sellingPrice,
  });

  Map<String, dynamic> toJson() {
    return {
      'selling_price': sellingPrice,
    };
  }
}

File: internal/domain/models/vehicle.go
package models

import (
	"time"
)

// VehicleType represents the vehicle_types table
type VehicleType struct {
	ID          int       `json:"id" db:"id"`
	Name        string    `json:"name" db:"name" validate:"required,max=100"`
	Description *string   `json:"description" db:"description"`
	CreatedAt   time.Time `json:"created_at" db:"created_at"`
}

// VehicleTypeCreateRequest for creating new vehicle type
type VehicleTypeCreateRequest struct {
	Name        string  `json:"name" validate:"required,max=100"`
	Description *string `json:"description"`
}

// VehicleTypeUpdateRequest for updating vehicle type
type VehicleTypeUpdateRequest struct {
	Name        *string `json:"name" validate:"omitempty,max=100"`
	Description *string `json:"description"`
}

// VehicleBrand represents the vehicle_brands table
type VehicleBrand struct {
	ID          int          `json:"id" db:"id"`
	Name        string       `json:"name" db:"name" validate:"required,max=100"`
	TypeID      int          `json:"type_id" db:"type_id" validate:"required"`
	CreatedAt   time.Time    `json:"created_at" db:"created_at"`
	VehicleType *VehicleType `json:"vehicle_type,omitempty"`
}

// VehicleBrandCreateRequest for creating new vehicle brand
type VehicleBrandCreateRequest struct {
	Name   string `json:"name" validate:"required,max=100"`
	TypeID int    `json:"type_id" validate:"required"`
}

// VehicleBrandUpdateRequest for updating vehicle brand
type VehicleBrandUpdateRequest struct {
	Name   *string `json:"name" validate:"omitempty,max=100"`
	TypeID *int    `json:"type_id"`
}

// SourceType enum
type SourceType string

const (
	SourceTypeCustomer SourceType = "customer"
	SourceTypeSupplier SourceType = "supplier"
)

// ConditionStatus enum
type ConditionStatus string

const (
	ConditionExcellent   ConditionStatus = "excellent"
	ConditionGood        ConditionStatus = "good"
	ConditionFair        ConditionStatus = "fair"
	ConditionPoor        ConditionStatus = "poor"
	ConditionNeedsRepair ConditionStatus = "needs_repair"
)

// VehicleStatus enum
type VehicleStatus string

const (
	VehicleStatusAvailable VehicleStatus = "available"
	VehicleStatusInRepair  VehicleStatus = "in_repair"
	VehicleStatusSold      VehicleStatus = "sold"
	VehicleStatusReserved  VehicleStatus = "reserved"
)

// Vehicle represents the vehicles table
type Vehicle struct {
	ID               int             `json:"id" db:"id"`
	Code             string          `json:"code" db:"code" validate:"required,max=50"`
	BrandID          int             `json:"brand_id" db:"brand_id" validate:"required"`
	Model            string          `json:"model" db:"model" validate:"required,max=100"`
	Year             int             `json:"year" db:"year" validate:"required,min=1980"`
	Color            *string         `json:"color" db:"color" validate:"omitempty,max=50"`
	EngineCapacity   *string         `json:"engine_capacity" db:"engine_capacity" validate:"omitempty,max=20"`
	FuelType         *string         `json:"fuel_type" db:"fuel_type" validate:"omitempty,max=20"`
	TransmissionType *string         `json:"transmission_type" db:"transmission_type" validate:"omitempty,max=20"`
	LicensePlate     *string         `json:"license_plate" db:"license_plate" validate:"omitempty,max=20"`
	ChassisNumber    *string         `json:"chassis_number" db:"chassis_number" validate:"omitempty,max=100"`
	EngineNumber     *string         `json:"engine_number" db:"engine_number" validate:"omitempty,max=100"`
	Odometer         int             `json:"odometer" db:"odometer"`
	SourceType       SourceType      `json:"source_type" db:"source_type" validate:"required"`
	SourceID         *int            `json:"source_id" db:"source_id"`
	PurchasePrice    float64         `json:"purchase_price" db:"purchase_price" validate:"required,min=0"`
	ConditionStatus  ConditionStatus `json:"condition_status" db:"condition_status" validate:"required"`
	Status           VehicleStatus   `json:"status" db:"status"`
	RepairCost       *float64        `json:"repair_cost" db:"repair_cost"`
	HPPPrice         *float64        `json:"hpp_price" db:"hpp_price"`
	SellingPrice     *float64        `json:"selling_price" db:"selling_price"`
	SoldPrice        *float64        `json:"sold_price" db:"sold_price"`
	SoldDate         *time.Time      `json:"sold_date" db:"sold_date"`
	Notes            *string         `json:"notes" db:"notes"`
	CreatedBy        int             `json:"created_by" db:"created_by" validate:"required"`
	CreatedAt        time.Time       `json:"created_at" db:"created_at"`
	UpdatedAt        time.Time       `json:"updated_at" db:"updated_at"`
	Brand            *VehicleBrand   `json:"brand,omitempty"`
	Creator          *User           `json:"creator,omitempty"`
	Photos           []VehiclePhoto  `json:"photos,omitempty"`
}

// VehiclePhoto represents the vehicle_photos table
type VehiclePhoto struct {
	ID        int       `json:"id" db:"id"`
	VehicleID int       `json:"vehicle_id" db:"vehicle_id" validate:"required"`
	PhotoPath string    `json:"photo_path" db:"photo_path" validate:"required,max=255"`
	IsPrimary bool      `json:"is_primary" db:"is_primary"`
	Caption   *string   `json:"caption" db:"caption" validate:"omitempty,max=255"`
	CreatedAt time.Time `json:"created_at" db:"created_at"`
}

// VehicleCreateRequest for creating new vehicle
type VehicleCreateRequest struct {
	Code             string          `json:"code" validate:"required,max=50"`
	BrandID          int             `json:"brand_id" validate:"required"`
	Model            string          `json:"model" validate:"required,max=100"`
	Year             int             `json:"year" validate:"required,min=1980"`
	Color            *string         `json:"color" validate:"omitempty,max=50"`
	EngineCapacity   *string         `json:"engine_capacity" validate:"omitempty,max=20"`
	FuelType         *string         `json:"fuel_type" validate:"omitempty,max=20"`
	TransmissionType *string         `json:"transmission_type" validate:"omitempty,max=20"`
	LicensePlate     *string         `json:"license_plate" validate:"omitempty,max=20"`
	ChassisNumber    *string         `json:"chassis_number" validate:"omitempty,max=100"`
	EngineNumber     *string         `json:"engine_number" validate:"omitempty,max=100"`
	Odometer         int             `json:"odometer"`
	SourceType       SourceType      `json:"source_type" validate:"required"`
	SourceID         *int            `json:"source_id"`
	PurchasePrice    float64         `json:"purchase_price" validate:"required,min=0"`
	ConditionStatus  ConditionStatus `json:"condition_status" validate:"required"`
	Notes            *string         `json:"notes"`
}

// VehicleUpdateRequest for updating vehicle
type VehicleUpdateRequest struct {
	Model            *string          `json:"model" validate:"omitempty,max=100"`
	Year             *int             `json:"year" validate:"omitempty,min=1980"`
	Color            *string          `json:"color" validate:"omitempty,max=50"`
	EngineCapacity   *string          `json:"engine_capacity" validate:"omitempty,max=20"`
	FuelType         *string          `json:"fuel_type" validate:"omitempty,max=20"`
	TransmissionType *string          `json:"transmission_type" validate:"omitempty,max=20"`
	LicensePlate     *string          `json:"license_plate" validate:"omitempty,max=20"`
	ChassisNumber    *string          `json:"chassis_number" validate:"omitempty,max=100"`
	EngineNumber     *string          `json:"engine_number" validate:"omitempty,max=100"`
	Odometer         *int             `json:"odometer"`
	ConditionStatus  *ConditionStatus `json:"condition_status"`
	Status           *VehicleStatus   `json:"status"`
	SellingPrice     *float64         `json:"selling_price" validate:"omitempty,min=0"`
	Notes            *string          `json:"notes"`
}

// VehicleSearchFilters represents the search filters for vehicles
type VehicleSearchFilters struct {
	BrandID     *int     `form:"brand_id" json:"brand_id,omitempty"`
	Model       string   `form:"model" json:"model,omitempty"`
	YearMin     *int     `form:"year_min" json:"year_min,omitempty"`
	YearMax     *int     `form:"year_max" json:"year_max,omitempty"`
	Color       string   `form:"color" json:"color,omitempty"`
	OdometerMin *int     `form:"odometer_min" json:"odometer_min,omitempty"`
	OdometerMax *int     `form:"odometer_max" json:"odometer_max,omitempty"`
	PriceMin    *float64 `form:"price_min" json:"price_min,omitempty"`
	PriceMax    *float64 `form:"price_max" json:"price_max,omitempty"`
	Status      string   `form:"status" json:"status,omitempty"`
}
